libs/core/langchain_core/indexing/schema_02.py

```diff
@@ -0,0 +1,265 @@
+    Dict,
+    Generic,
+    Iterable,
+    List,
+    NotRequired,
+    Optional,
+    Sequence,
+    TypedDict,
+    TypeVar,
+    Union,
+)
+
+from langchain_core.load.serializable import Serializable
+from langchain_core.runnables import RunnableConfig
+
+
+class IndexableContent(
+    ABC, Serializable
+):  # Introduce new base class for content. Name OK?
+    id: str
+    metadata: Optional[dict]  # In a more extreme form we could remove
+
+
+class Document(IndexableContent):  # Re-use existing Document model in langchain_core
+    content: str
+
+
+class Blob(IndexableContent):  # Re-use existing Blob model in langchain_core
+    """Use for images / audio / video"""
+
+    data: bytes
+    encoding: str  # base64, utf-8, URI, etc.
+    mimetype: str  # image/png, text/plain, etc.
+
+
+# This is likely not needed during indexing right now
+# class MultiContent(Content):
+#     """Group multiple pieces of content together"""
+#     contents: List[Content]
+#     metadata: dict
+
+
+T = TypeVar("T", bound=IndexableContent)
+
+
+class UpsertResponse(TypedDict):
+    """An indexing result."""
+
+    succeeded: Sequence[str]
+    failed: Sequence[str]
+
+
+class DeleteResponse(TypedDict):
+    """A response to a delete request."""
+
+    num_deleted: NotRequired[int]
+    num_failed: NotRequired[int]
+    succeeded: NotRequired[Sequence[str]]
+    failed: NotRequired[Sequence[str]]
+
+
+class Sort(TypedDict):
+    """A sort object."""
+
+    field: str
+    ascending: NotRequired[bool]  # Assume asc=True
+
+
+C = TypeVar("C")
+
+
+class BaseIndex(Generic[T]):
+    """An index represent a collection of items that can be queried.
+
+    The index is responsible for:
+
+    1. Storing the content
+    2. Allowing retrieval of the content by id
+    3. Supporting queries against the metadata associated with the content
+
+    The index is **NOT** responsible for:
+
+    1. Support search queries against the content.
+
+    The types of supported queries depends on the implementation (e.g., a vectorstore index)
+    """
+
+    @abc.abstractmethod
+    def upsert(self, data: Iterable[T], /, **kwargs: Any) -> Iterable[UpsertResponse]:
+        """Upsert a stream of data by id."""
+
+    @abc.abstractmethod
+    def get_by_ids(
+        self,
+        ids: Sequence[str],
+        /,
+    ) -> List[T]:
+        """Get items by id."""
+
+    @abc.abstractmethod
+    def delete_by_ids(
+        self,
+        ids: Sequence[str],
+        /,
+        **kwargs: Any,
+    ) -> DeleteResponse:
+        """Delete an item by id."""
+
+    # Delete and get are part of the READ/WRITE interface.
+    # They do not take advantage of indexes on the content.
+    # However, all the indexers ARE assumed to have the capability to index
+    # on metadata if they implement the delete_by_query and get_by_query methods.
+    @abc.abstractmethod
+    def get_by_query(
+        self,
+        *,
+        filter: Optional[Union[List[Dict[str, Any], Dict[str, Any]]]] = None,
```

Comment:
Alternatives:

1. Use a generic for the entire `Query`


```
class Indexer(Generic[Q]):
  def get_by_query(
     self, query: Q
  ):
```

This seems to be the most bullet proof way to handle different types of queries.

2. Use a generic on the filter

```python
  @abc.abstractmethod
    def get_by_query(
        self,
        *,
        filter: Optional[F] = None,
```

---

In each case the goal would be to allow implementations to support both a LangChain representation for filters/queries  as well supporting implementation native functionality (which is usually much richer than what LangChain can accommodate)

---

The main concern is for end-user UX and having to input dictionaries instead of relying on the function signature more directly